_posts/2017-03-30-storage-formats-quickstep.markdown

```diff
@@ -0,0 +1,140 @@
+
+{% highlight sql %}
+INSERT into employee values (101, 30, Jennifer);
+INSERT into employee values (102, 25, Jim);
+INSERT into employee values (103, 35, David);
+{% endhighlight %}
+
+Conceptually, our table looks as follows:
+
+| ID 	| Age 	|   Name   	|
+|:--:	|:---:	|:--------:	|
+|  101 	|  30 	| Jennifer 	|
+|  102 	|  25 	|    Jim   	|
+|  103 	|  35 	|   David  	|
+
+Next, let us dig into understanding the different ways in which this table can be stored in memory. 
+
+<h2>Row Store</h2>
+
+The first storage format we will look at is *row store*, which is a popular format, and one that is intuitive too.
+
+The figure below represents how a row store format for our toy example looks like in memory. The offsets indicate how far is the memory location from the beginning of the block. 
+
+![Row Store](../assets/storage-format-row-store.jpg)
+
+In the row store format, a tuple is stored in the table definition order. In memory, the ID (101) of the first tuple is stored at some address, and it is followed by the age field for that tuple (30), and then the name field ("Jennifer"). The second tuple is stored in memory sequentially after the first tuple, and so on.
+
+<h2>Column Store</h2>
+
+Let's now take a look at the *column store* format. The figure below depicts a column store layout for our toy example. 
+
+![Column Store](../assets/storage-format-column-store.jpg)
+
+In the column store format, the values from the same column are stored together. The order in which the values from a column are stored remains the same for all the columns. Notice the ID values 101, 102, 103 are stored contiguously. The corresponding Age fields 30, 25 and 35 are stored contiguously and likewise for the Name column.
+
+<h2>Compression</h2>
+
+Compression is a standard technique to reduce the storage footprint. There is a large body of work on various compression techniques, which we won't cover here. Let's look at how compression can be applied in our toy example. If we look at the ID column, it has three values 101, 102 and 103. If they are stored as regular integers, each of them will occupy 4 bytes of memory. Interestingly, we can reduce the memory consumed in storing these three values. Observe that these three values are very close to 100. If we remember the base value 100 and only record the difference of each value from this base value (of 100), then we only need to store the values 1, 2, and 3. To store these three *deltas*, we don't even need a 4 byte integer, a 1-byte value is enough.
+
+Putting it all together, the picture below shows the compressed column store format in which the compression is applied on the ID column. 
+
+![Column Store with Compression](../assets/storage-format-compressed-column-store.jpg)
```

Comment:
My guess is that preferred way to do this is with an absolute path: `/assets/storage-format-compressed-column-store.jpg`